test/integration/consul-container/test/multiport/explicit_destination_test.go

```diff
@@ -0,0 +1,248 @@
+//   - Make sure a request from static-client to the virtual address (<svc>.virtual.consul) returns a
+//     response from the upstream.
+func TestMultiportService_Explicit(t *testing.T) {
+	t.Parallel()
+
+	cluster := createCluster(t)
+	followers, err := cluster.Followers()
+	require.NoError(t, err)
+	client := pbresource.NewResourceServiceClient(followers[0].GetGRPCConn())
+	resourceClient := rtest.NewClient(client)
+
+	serverIP := cluster.Agents[1].GetIP()
+	clientIP := cluster.Agents[2].GetIP()
+
+	serverService := createServerServicesAndWorkloads(t, resourceClient, serverIP)
+	createClientResources(t, resourceClient, serverService, clientIP)
+
+	_, clientDataplane := createServices(t, cluster)
+
+	_, port := clientDataplane.GetAddr()
+
+	assertDataplaneContainerState(t, clientDataplane, "running")
+	libassert.HTTPServiceEchoes(t, "localhost", port, "")
+	libassert.AssertFortioName(t, fmt.Sprintf("http://localhost:%d", port), "static-server", "")
+}
+
+// createServices creates the static-client and static-server services with
+// transparent proxy enabled. It returns a Service for the static-client.
+func createServices(t *testing.T, cluster *libcluster.Cluster) (*libcluster.ConsulDataplaneContainer, *libcluster.ConsulDataplaneContainer) {
+	n1 := cluster.Agents[1]
+
+	// Create a service and dataplane
+	serverDataplane, err := createServiceAndDataplane(t, n1, "static-server-workload", "static-server", 8080, 8079, []int{})
+	require.NoError(t, err)
+
+	n2 := cluster.Agents[2]
+	// Create a service and dataplane
+	clientDataplane, err := createServiceAndDataplane(t, n2, "static-client-workload", "static-client", 8080, 8079, []int{libcluster.ServiceUpstreamLocalBindPort})
+	require.NoError(t, err)
+
+	return serverDataplane, clientDataplane
+}
+
+func createServiceAndDataplane(t *testing.T, node libcluster.Agent, proxyID, serviceName string, httpPort, grpcPort int, serviceBindPorts []int) (*libcluster.ConsulDataplaneContainer, error) {
+	// Do some trickery to ensure that partial completion is correctly torn
+	// down, but successful execution is not.
+	var deferClean utils.ResettableDefer
+	defer deferClean.Execute()
+
+	// Create a service and proxy instance
+	svc, err := libservice.NewExampleService(context.Background(), serviceName, httpPort, grpcPort, node)
+	if err != nil {
+		return nil, err
+	}
+	deferClean.Add(func() {
+		_ = svc.Terminate()
+	})
+
+	// Create Consul Dataplane
+	dp, err := libcluster.NewConsulDataplane(context.Background(), proxyID, "0.0.0.0", 8502, serviceBindPorts, node)
+	require.NoError(t, err)
+	deferClean.Add(func() {
+		_ = dp.Terminate()
+	})
+
+	// disable cleanup functions now that we have an object with a Terminate() function
+	deferClean.Reset()
+
+	return dp, nil
+}
+
+func createServerServicesAndWorkloads(t *testing.T, resourceClient *rtest.Client, ipAddress string) *pbresource.Resource {
+	serverService := rtest.ResourceID(&pbresource.ID{
+		Name: "static-server-service",
+		Type: &pbresource.Type{
```

Comment:
It looks like these are available in proto-public now https://github.com/hashicorp/consul/blob/main/proto-public/pbcatalog/v2beta1/resource_types.gen.go#L55